File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/drive/advanced/subsystems/DistanceTrigger.java
package org.firstinspires.ftc.teamcode.drive.advanced.subsystems;

import com.qualcomm.robotcore.hardware.ColorRangeSensor;

import org.firstinspires.ftc.robotcore.external.navigation.DistanceUnit;
import org.firstinspires.ftc.teamcode.drive.galahlib.ButtonPressed;

public class DistanceTrigger {
    public boolean val;
    public ButtonPressed onChange;
    public ColorRangeSensor sensor;
    public double triggerDistance;

    public DistanceTrigger(double triggerDistance, ColorRangeSensor sensor, ButtonPressed onChange) {
        this.onChange = onChange;
        this.sensor = sensor;
        this.triggerDistance = triggerDistance;
    }

    public DistanceTrigger(double triggerDistance, ColorRangeSensor sensor) {
        this.sensor = sensor;
        this.triggerDistance = triggerDistance;
    }

    public void update() {
        boolean oldVal = val;
        val = sensor.getDistance(DistanceUnit.CM) < triggerDistance;

        if (oldVal != val && onChange != null) {
            onChange.call(val);
        }
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/galahlib/mechanisms/DistanceTrigger.kt
package org.firstinspires.ftc.teamcode.galahlib.mechanisms

import com.acmerobotics.dashboard.telemetry.TelemetryPacket
import com.acmerobotics.roadrunner.Action
import com.qualcomm.robotcore.hardware.DistanceSensor
import org.firstinspires.ftc.robotcore.external.navigation.DistanceUnit
import org.firstinspires.ftc.teamcode.staticData.PoseStorage

class DistanceTrigger(val sensor: DistanceSensor) {
    @JvmOverloads
    fun closerThan(distance: Double, units: DistanceUnit = DistanceUnit.MM): Action {
        return object : Action {
            override fun run(p: TelemetryPacket): Boolean {
                return sensor.getDistance(units) > distance && !PoseStorage.shouldHallucinate
            }
        }
    }

    @JvmOverloads
    fun furtherThan(distance: Double, units: DistanceUnit = DistanceUnit.MM): Action {
        return object : Action {
            override fun run(p: TelemetryPacket): Boolean {
                return sensor.getDistance(units) < distance && !PoseStorage.shouldHallucinate
            }
        }
    }
}